official/projects/lra/linformer_encoder_block.py

```diff
@@ -0,0 +1,446 @@
+    self._use_bias = use_bias
+    self._norm_first = norm_first
+    self._norm_epsilon = norm_epsilon
+    self._inner_dropout = inner_dropout
+    self._use_query_residual = use_query_residual
+    self._key_dim = key_dim
+    self._value_dim = value_dim
+    self._output_last_dim = output_last_dim
+    self._diff_q_kv_att_layer_norm = diff_q_kv_att_layer_norm
+    self._return_attention_scores = return_attention_scores
+    if attention_initializer:
+      self._attention_initializer = tf.keras.initializers.get(
+          attention_initializer)
+    else:
+      self._attention_initializer = tf_utils.clone_initializer(
+          self._kernel_initializer)
+    self._attention_axes = attention_axes
+
+    if self._diff_q_kv_att_layer_norm and not self._norm_first:
+      raise ValueError("Setting `diff_q_and_kv_attention_layer_norm` to True"
+                       "when `norm_first` is False is invalid.")
+
+  def build(self, input_shape):
+    if isinstance(input_shape, tf.TensorShape):
+      input_tensor_shape = input_shape
+    elif isinstance(input_shape, (list, tuple)):
+      input_tensor_shape = tf.TensorShape(input_shape[0])
+    else:
+      raise ValueError(
+          "The type of input shape argument is not supported, got: %s" %
+          type(input_shape))
+    einsum_equation = "abc,cd->abd"
+    if len(input_tensor_shape.as_list()) > 3:
+      einsum_equation = "...bc,cd->...bd"
+    hidden_size = input_tensor_shape[-1]
+    if hidden_size % self._num_heads != 0:
+      logging.warning(
+          "The input size (%d) is not a multiple of the number of attention "
+          "heads (%d)", hidden_size, self._num_heads)
+    if self._key_dim is None:
+      self._key_dim = int(hidden_size // self._num_heads)
+    if self._output_last_dim is None:
+      last_output_shape = hidden_size
+    else:
+      last_output_shape = self._output_last_dim
+
+    common_kwargs = dict(bias_regularizer=self._bias_regularizer,
+                         activity_regularizer=self._activity_regularizer,
+                         kernel_constraint=self._kernel_constraint,
+                         bias_constraint=self._bias_constraint)
+    self._key_projection = tf.keras.layers.Dense(
+        self._low_rank_features,
+        activation=None,
+        use_bias=False,
+        kernel_initializer=tf_utils.clone_initializer(self._kernel_initializer),
+        bias_initializer=tf_utils.clone_initializer(self._bias_initializer),
+        name="key_low_rank_projection",
+        **common_kwargs)
+    self._value_projection = tf.keras.layers.Dense(
+        self._low_rank_features,
+        activation=None,
+        use_bias=False,
+        kernel_initializer=tf_utils.clone_initializer(self._kernel_initializer),
+        bias_initializer=tf_utils.clone_initializer(self._bias_initializer),
+        name="value_low_rank_projection",
+        **common_kwargs)
+    self._attention_layer = tf.keras.layers.MultiHeadAttention(
```

Comment:
it is still using the MHA?
May you tell the difference of lin former?

Reply:
Yes, it has an additional dimension reduction at the front